modules/nextflow/src/main/groovy/nextflow/scm/AzureDevOpsRepositoryProvider.groovy

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2020, Seqera Labs
+ * Copyright 2013-2019, Centre for Genomic Regulation (CRG)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nextflow.scm
+
+import groovy.transform.CompileDynamic
+import groovy.transform.CompileStatic
+import groovy.transform.Memoized
+
+/**
+ * Implements a repository provider for Azure DevOps service
+ *
+ * @author Tobias Neumann <tobias.neumann.at@gmail.com>
+ */
+@CompileStatic
+final class AzureDevOpsRepositoryProvider extends RepositoryProvider {
+
+    private String user
+    private String repo
+
+    AzureDevOpsRepositoryProvider(String project, ProviderConfig config=null) {
+        this.project = project
+        this.user = this.project.tokenize('/').first()
+        this.repo = this.project.tokenize('/').last()
+        this.config = config ?: new ProviderConfig('azurerepos')
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    String getName() { "Azure Repos" }
+
+    /** {@inheritDoc} */
+    @Override
+    String getEndpointUrl() {
+        "${config.endpoint}/${project}/_apis/git/repositories/${repo}"
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    String getContentUrl( String path ) {
+        "${config.endpoint}/${project}/_apis/git/repositories/${repo}/items?download=false&includeContent=true&includeContentMetadata=false&api-version=6.0&\$format=json&path=$path"
+    }
+
+    @Override
+    @CompileDynamic
+    List<BranchInfo> getBranches() {
+        // https://docs.microsoft.com/en-us/rest/api/azure/devops/git/refs/list?view=azure-devops-rest-6.0#refs-heads
+        final url = "${config.endpoint}/${project}/_apis/git/repositories/${repo}/refs?filter=heads&api-version=6.0"
+        this.<BranchInfo>invokeAndResponseWithPaging(url, { Map branch -> new BranchInfo(branch.name as String, branch.id as String) })
+    }
+
+    @Override
+    @CompileDynamic
+    @Memoized
+    List<TagInfo> getTags() {
+        // https://docs.microsoft.com/en-us/rest/api/azure/devops/git/refs/list?view=azure-devops-rest-6.0#refs-tags
+        final url = "${config.endpoint}/${project}/_apis/git/repositories/${repo}/refs?filter=tags&api-version=6.0"
```

Comment:
Same here, I would expect Get Tags 

https://docs.microsoft.com/en-us/rest/api/azure/devops/build/tags/get%20tags?view=azure-devops-rest-6.0

Reply:
See above

Reply:
it's a bit counterintuitive, but I guess you are right. Need to make a last check.

Reply:
@pditommaso This week maybe? Let me know if I can help with anything

Reply:
There were some problems with the tag and branch parsing. What it's missing I was unable to find is the how to manage the URL pagination. Any idea about that?

Reply:
Can u show me an example - when I checked on my repo, there was no pagination required as it was all in one large json list.

Reply:
It depends on the API. Github uses [`page` parameters](https://docs.github.com/en/free-pro-team@latest/rest/guides/traversing-with-pagination). Other API provides the next page URL in the response. It should be specified in the docs.

Reply:
Yeah I read up on this and they have an `$top` parameter in the URL which per default lists 1000 entries and also supply a pagination token if there is > 1000 entries.

So I think for our purposes this should not be required - also failed to retrieve a paginationToken if I hardcoded `$top=1` in the URL.

Reply:
This is my example json output for the following query:

```
https://dev.azure.com/quantrocode/slamseq/_apis/git/repositories/slamseq/refs?filter=heads&api-version=6.0
```

```
{"value":
[{"name":"refs/heads/TEMPLATE","objectId":"2244c689382d0045054b3f58a7debfd10ab12d08","creator":{"displayName":"Tobias Neumann","url":"https://spsprodweu1.vssps.visualstudio.com/Ae124d364-fa73-4713-93c7-c540f24c8553/_apis/Identities/c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","_links":{"avatar":{"href":"https://dev.azure.com/quantrocode/_apis/GraphProfile/MemberAvatars/aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"}},"id":"c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","uniqueName":"tobias.neumann@quantro-tx.com","imageUrl":"https://dev.azure.com/quantrocode/_api/_common/identityImage?id=c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","descriptor":"aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"},"url":"https://dev.azure.com/quantrocode/5f2cadab-32ef-4ac9-9fe5-28252f9a669f/_apis/git/repositories/cb5bd752-6d3c-459c-be6d-deb87f15c6ae/refs?filter=heads%2FTEMPLATE"},
{"name":"refs/heads/dev","objectId":"94804400ce86a89171ab09428ab93fed6c560e5f","creator":{"displayName":"Tobias Neumann","url":"https://spsprodweu1.vssps.visualstudio.com/Ae124d364-fa73-4713-93c7-c540f24c8553/_apis/Identities/c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","_links":{"avatar":{"href":"https://dev.azure.com/quantrocode/_apis/GraphProfile/MemberAvatars/aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"}},"id":"c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","uniqueName":"tobias.neumann@quantro-tx.com","imageUrl":"https://dev.azure.com/quantrocode/_api/_common/identityImage?id=c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","descriptor":"aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"},"url":"https://dev.azure.com/quantrocode/5f2cadab-32ef-4ac9-9fe5-28252f9a669f/_apis/git/repositories/cb5bd752-6d3c-459c-be6d-deb87f15c6ae/refs?filter=heads%2Fdev"},
{"name":"refs/heads/master","objectId":"081d363591ca8b1adb7325099f0672e6f467c320","creator":{"displayName":"Tobias Neumann","url":"https://spsprodweu1.vssps.visualstudio.com/Ae124d364-fa73-4713-93c7-c540f24c8553/_apis/Identities/c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","_links":{"avatar":{"href":"https://dev.azure.com/quantrocode/_apis/GraphProfile/MemberAvatars/aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"}},"id":"c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","uniqueName":"tobias.neumann@quantro-tx.com","imageUrl":"https://dev.azure.com/quantrocode/_api/_common/identityImage?id=c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","descriptor":"aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"},"url":"https://dev.azure.com/quantrocode/5f2cadab-32ef-4ac9-9fe5-28252f9a669f/_apis/git/repositories/cb5bd752-6d3c-459c-be6d-deb87f15c6ae/refs?filter=heads%2Fmaster"},
{"name":"refs/heads/quantro-dev","objectId":"6e2ede3d72c591e956f921ff0b2fcef017af1964","creator":{"displayName":"Tobias Neumann","url":"https://spsprodweu1.vssps.visualstudio.com/Ae124d364-fa73-4713-93c7-c540f24c8553/_apis/Identities/c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","_links":{"avatar":{"href":"https://dev.azure.com/quantrocode/_apis/GraphProfile/MemberAvatars/aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"}},"id":"c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","uniqueName":"tobias.neumann@quantro-tx.com","imageUrl":"https://dev.azure.com/quantrocode/_api/_common/identityImage?id=c4307a7c-514e-6e2b-b0a4-ce6b92b21c63","descriptor":"aad.YzQzMDdhN2MtNTE0ZS03ZTJiLWIwYTQtY2U2YjkyYjIxYzYz"},"url":"https://dev.azure.com/quantrocode/5f2cadab-32ef-4ac9-9fe5-28252f9a669f/_apis/git/repositories/cb5bd752-6d3c-459c-be6d-deb87f15c6ae/refs?filter=heads%2Fquantro-dev"}],
"count":4}
```

Reply:
When are these methods called anyways? I have a hard time to deduce this from the code. Once you pull it in all looked fine, but the suspect they use a regular `git branch` command on it then.

Reply:
Could you please provide the link to the docs where describe the pagination?

Reply:
Here is the link to the docs where the `$top` and `continuationToken` is described:

https://docs.microsoft.com/en-us/rest/api/azure/devops/git/refs/list?view=azure-devops-rest-6.0

And here is a similar page with more details on the `continuationToken` I found and some discussion on StackOverflow.

https://docs.microsoft.com/en-us/rest/api/azure/devops/test/configurations/list?view=azure-devops-rest-5.0

https://stackoverflow.com/questions/63986932/azure-devops-rest-api-pagination-and-rate-limit

Reply:
The methods are called when resolving the info for a remote repository.  However, `$top` default to 100, not 1000. 

If think it's needed to add the support for the `continuationToken`, tho it's not clear to me how to retrieve it.

Reply:
Ok so `$top` can also be just set to 1000. If that is enough I would set it so.

If you insist on the `continuationToken`, can u just send me an example Nextflow call where the methods are actually called so I can properly debug it?

Reply:
You can this snippet in the azure repository test to check tags and a similar one for branches 

```
    def 'should fetch repo tags'() {

        given:
        def token = System.getenv('NXF_AZURE_REPOS_TOKEN')
        def config = new ProviderConfig('azurerepos').setAuth(token)

        when:
        def repo = new AzureRepositoryProvider('paolo0758/nf-azure-repo', config)
        def result = repo.getTags()
        then:
        result == [
                new RepositoryProvider.TagInfo('v1.0', '769afa02858ef904756eb5b9a9a04ce8ef869f42'),
                new RepositoryProvider.TagInfo('v2.0', '1f73719675872801cc32fa7fc4ac7f2c04320714')]

    }
```

using your own azure repo of course.

Reply:
Looks like reading the credentials from the `scm` config did not work. How would I set them for the testcase?

```
providers {

    azurerepos {
        user = 'XXXXXX'
        password = 'XXXXX'
    }

}
```

Reply:
the first two lines do that token is `"${user}:${password}"`

Reply:
Hi again,

so I found out how the pagination works - maybe you can quickly guide on how to implement it. So if there are more pages the http request returns a `x-ms-continuationtoken` in the header. While this is set, the next page can be retrieved by adding this via `continuationToken` to the followup request until no more `x-ms-continuationtoken` is returned. So in the end it will be a loop through all pages, but these are not explicitely listed in the json content but only visible via this `x-ms-continuationtoken` in the response header. How can I best retrieve the response header variables within the current framework?

Reply:
There not could be a more weird approach. What you can try to do is to add class field `continuationToken` in the Azure provider. Then set if overriding the [checkResponse](https://github.com/nextflow-io/nextflow/blob/e94e6877c2dc9bed983e873b62e302a08843c891/modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy#L157) method, which first check for that header, than continue with the `super.checkResponse(connection)` 

Then use the  `continuationToken` in the iteration made in the `invokeAndResponseWithPaging` method

Reply:
I implemented your suggestions and tested them with manually setting `$top` to 1. I retreived all follow-up entries, so looks good to me now.